src/main/java/backend/Logic.java

```diff
@@ -247,13 +209,204 @@ public Model getRepo(String repoId) {
                     replaceIssueLabelsUI(issue, originalLabels);
                     logger.error(e.getLocalizedMessage(), e);
                     return false;
-        });
+                });
     }
 
     public void replaceIssueLabelsUI(TurboIssue issue, List<String> labels) {
         logger.info(HTLog.format(issue.getRepoId(), "Applying labels " + labels + " to " + issue + " in UI"));
         issue.setLabels(labels);
-        updateUIAndShow();
+        refreshUI();
     }
 
+    /**
+     * Determines data to be sent to the GUI to refresh the entire GUI with the current model in Logic,
+     * and then sends the data to the GUI.
+     */
+    private void refreshUI() {
+        filterSortRefresh(getAllUIFilters());
+    }
+
+    /**
+     * Given a list of filter expressions, dispatch metadata update if needed and then process them to return
+     * a map of filtered and sorted issues corresponding to each filter expression, based on the most recent data
+     * from the repository source.
+     *
+     * @param filterTexts Filter expressions to process
+     */
+    public void filterSortRefresh(List<FilterExpression> filterTexts) {
+        // Open specified repos
+        openRepositoriesInFilters(filterTexts);
+
+        // First filter, for issues requiring a metadata update.
+        Map<String, List<TurboIssue>> toUpdate = tallyMetadataUpdate(filterTexts);
+
+        if (toUpdate.size() > 0) {
+            // If there are issues requiring metadata update, we dispatch the metadata requests...
+            ArrayList<CompletableFuture<Boolean>> metadataRetrievalTasks = new ArrayList<>();
+            toUpdate.forEach((repoId, issues) -> {
+                metadataRetrievalTasks.add(getIssueMetadata(repoId, issues));
+            });
+            // ...and then wait for all of them to complete.
+            CompletableFuture<Void> allCompletables = CompletableFuture.allOf(
```

Comment:
Ah yup, I didn't think it'd be necessary to use `Futures.sequence` since we aren't currently using the return values from `getIssueMetadata`. But it makes sense to prep it for a future PR.